File: checkout-ruby/lib/checkout.rb
require 'pry'

class Checkout
  attr_accessor :cart_promotional_rule_list, :item_promotional_rule_map

  def initialize(promotional_rule_list)
    @cart_promotional_rule_list = PromotionalRuleHandlerFactory.get_cart_promotional_list(promotional_rule_list)
    @item_promotional_rule_map = PromotionalRuleHandlerFactory.get_item_promotional_map(promotional_rule_list)
    @bucket_map = {}
  end

  def scan(item)
    item_product_code = item.product_code
    checkout_item = bucket_map[item_product_code]
    if checkout_item.nil?
      rule = item_promotional_rule_map[item_product_code]
      checkout_item = CheckoutItem.new(item, rule)
      bucket_map[item_product_code] = checkout_item
    end
    checkout_item.add_item()
  end

  def total
    total_price = 0.0
    bucket_map.values.each do |item|
      total_price += item.get_price
    end

    cart_promotional_rule_list.each do |cart_promo_rule|
      if cart_promo_rule.is_eligible_for_discount?(total_price)
        total_price -= cart_promo_rule.value_type.get_discounted_price(total_price)
      end
    end
    total_price
  end

  private
  attr_accessor :bucket_map
end

File: src/main/java/com/vinterior/checkout/models/Checkout.java
package com.vinterior.checkout.models;


import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class Checkout {

    private List<PromotionalRule> cartPromotionalRuleList;
    private Map<Integer, PromotionalRule> itemPromotionalRuleMap;

    private Map<Integer, CheckoutItem> bucketMap;

    public Checkout(List<PromotionalRule> promotionalRuleList) {
        cartPromotionalRuleList = PromotionalRuleHandlerFactory.getCartPromoHandlers(promotionalRuleList);
        itemPromotionalRuleMap = PromotionalRuleHandlerFactory.getItemPromoHandlers(promotionalRuleList);
        bucketMap = new HashMap<>();
    }


    /**
     * Add item in bucket, increasing count
     * @param item
     */
    public void scan(Item item) {
        CheckoutItem checkoutItem = bucketMap.get(item.getProductCode());
        if(null == checkoutItem) {
            PromotionalRule rule = itemPromotionalRuleMap.get(item.getProductCode());
            checkoutItem = new CheckoutItem(item, rule);
            bucketMap.put(item.getProductCode(), checkoutItem);
        }
        checkoutItem.addItem();

    }

    public double getTotalPrice() {
        double totalPrice = 0;
        for(CheckoutItem item: bucketMap.values()) {
            totalPrice += item.getPrice();
        }

        for(PromotionalRule cartPromoRule: cartPromotionalRuleList) {
            if(cartPromoRule.isEligibleForDiscount(totalPrice)) {
                totalPrice -= cartPromoRule.getValueType().getDiscountedPrice(totalPrice);
            }
        }
        return totalPrice;
    }
}
